Compiler hangs with uninitialized variable containing a struct that contains itself

The following code causes rustc to hang when built with `cargo run --release`

```rust
trait Apply {
    type Output<T>;
}
struct Identity;
impl Apply for Identity {
    type Output<T> = T;
}

struct Thing<A: Apply>(A::Output<Self>);

/*
Effectively:
struct Thing_Identity(Thing_Identity);
*/

fn foo<A: Apply>() {
    let _x: Thing<A>;
}

fn main() {
    foo::<Identity>();
}
```

Compiler output before the hang:

```
   Compiling playground v0.0.1 (/playground)
error[E0391]: cycle detected when computing layout of `Thing<Identity>`
   |
   = note: ...which requires computing layout of `<Identity as Apply>::Output<Thing<Identity>>`...
   = note: ...which again requires computing layout of `Thing<Identity>`, completing the cycle
note: cycle used when optimizing MIR for `main`
  --> src/main.rs:20:1
   |
20 | fn main() {
   | ^^^^^^^^^
   = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
```

See also #152885

### Meta

Reproducible on the playground with version `1.95.0-nightly (2026-02-24 859951e3c7c9d0322c39)`

--- a/tests/ui/layout/uninitialized-gat-projection-cycle-issue-153205.rs
+++ b/tests/ui/layout/uninitialized-gat-projection-cycle-issue-153205.rs
@@ -1,0 +1,27 @@
+//! Regression test for https://github.com/rust-lang/rust/issues/153205.
+//! An uninitialized recursive GAT projection used to hang during optimized MIR processing
+//! instead of reporting the layout cycle.
+
+//@ build-fail
+//@ compile-flags: -O
+
+trait Apply {
+    type Output<T>;
+}
+
+struct Identity;
+
+impl Apply for Identity {
+    type Output<T> = T;
+}
+
+struct Thing<A: Apply>(A::Output<Self>);
+//~^ ERROR cycle detected when computing layout of `Thing<Identity>`
+
+fn foo<A: Apply>() {
+    let _x: Thing<A>;
+}
+
+fn main() {
+    foo::<Identity>();
+}

--- a/tests/ui/layout/uninitialized-gat-projection-cycle-issue-153205.stderr
+++ b/tests/ui/layout/uninitialized-gat-projection-cycle-issue-153205.stderr
@@ -1,0 +1,22 @@
+error[E0391]: cycle detected when computing layout of `Thing<Identity>`
+  --> $DIR/uninitialized-gat-projection-cycle-issue-153205.rs:18:1
+   |
+LL | struct Thing<A: Apply>(A::Output<Self>);
+   | ^^^^^^^^^^^^^^^^^^^^^^
+   |
+note: ...which requires computing layout of `<Identity as Apply>::Output<Thing<Identity>>`...
+  --> $DIR/uninitialized-gat-projection-cycle-issue-153205.rs:9:5
+   |
+LL |     type Output<T>;
+   |     ^^^^^^^^^^^^^^
+   = note: ...which again requires computing layout of `Thing<Identity>`, completing the cycle
+note: cycle used when optimizing MIR for `main`
+  --> $DIR/uninitialized-gat-projection-cycle-issue-153205.rs:25:1
+   |
+LL | fn main() {
+   | ^^^^^^^^^
+   = note: for more information, see <https://rustc-dev-guide.rust-lang.org/overview.html#queries> and <https://rustc-dev-guide.rust-lang.org/query.html>
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0391`.

